lib/spack/spack/hooks/absolutify_elf_sonames.py

```diff
@@ -0,0 +1,291 @@
+# Copyright 2013-2022 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+import os
+import struct
+import sys
+
+import llnl.util.tty as tty
+from llnl.util.filesystem import BaseDirectoryVisitor, visit_directory_tree
+from llnl.util.lang import elide_list
+
+import spack.bootstrap
+import spack.relocate
+from spack.util.executable import Executable
+
+
+class ELF:
+    ELFCLASS32 = 1
+    ELFCLASS64 = 2
+    ELFDATA2LSB = 1
+    ELFDATA2MSB = 2
+    ET_DYN = 3
+    PT_DYNAMIC = 2
+    PT_INTERP = 3
+    DT_NULL = 0
+    DT_SONAME = 14
+
+
+def get_byte_at(byte_array, idx):
+    if sys.version_info[0] < 3:
+        return ord(byte_array[idx])
+    return byte_array[idx]
+
+
+def is_probably_a_shared_library_elf_file(f):
+    # Read the 32/64 bit class independent part of the header and validate
```

Comment:
Can most of what is in this file be moved to `spack.util.elf`, and can the hook be kept as a fairly minimal thing that just calls functions from that?  I think this code is going to end up being useful for binary relocation.